kani-compiler/src/kani_middle/attributes.rs

```diff
@@ -2,33 +2,278 @@
+    let parent_id = tcx.hir().get_parent_item(hir_id);
+    let parent_def_id = parent_id.to_def_id();
+
+    is_test_harness_description(tcx, parent_def_id)
+}
+
+/// Extract all Kani attributes for a given `def_id` if any exists.
+pub fn extract_harness_attributes(tcx: TyCtxt, def_id: DefId) -> Option<HarnessAttributes> {
+    let attributes = extract_kani_attributes(tcx, def_id);
+    trace!(?def_id, ?attributes, "extract_harness_attributes");
+    if attributes.contains_key(&KaniAttributeKind::Proof) {
+        Some(attributes.into_iter().fold(
+            HarnessAttributes::default(),
+            |mut harness, (kind, attributes)| {
+                match kind {
+                    KaniAttributeKind::Solver => {
+                        // Make sure the solver is not already set
+                        harness.solver = parse_solver(tcx, expect_single(tcx, kind, &attributes));
+                    }
+                    KaniAttributeKind::Stub => {
+                        harness.stubs = parse_stubs(tcx, attributes);
+                    }
+                    KaniAttributeKind::Unwind => {
+                        harness.unwind_value =
+                            parse_unwind(tcx, expect_single(tcx, kind, &attributes))
+                    }
+                    KaniAttributeKind::Proof => harness.proof = true,
+                };
+                harness
+            },
+        ))
+    } else {
+        None
+    }
+}
+
+fn expect_single<'a>(
+    tcx: TyCtxt,
+    kind: KaniAttributeKind,
+    attributes: &'a Vec<&'a Attribute>,
+) -> &'a Attribute {
+    let attr = attributes
+        .first()
+        .expect(&format!("expected at least one attribute {} in {attributes:?}", kind.as_ref()));
+    if attributes.len() > 1 {
+        tcx.sess.span_err(
+            attr.span,
+            &format!("only one '#[kani::{}]' attribute is allowed per harness", kind.as_ref()),
+        );
+    }
+    attr
+}
+
+/// Check that if an item is tagged with a proof_attribute, it is a valid harness.
+fn check_proof_attribute(tcx: TyCtxt, def_id: DefId, proof_attributes: &Vec<&Attribute>) {
+    assert!(!proof_attributes.is_empty());
+    let span = proof_attributes.first().unwrap().span;
+    if proof_attributes.len() > 1 {
+        tcx.sess.span_warn(proof_attributes[0].span, "duplicate attribute");
+    }
+
+    if tcx.def_kind(def_id) != DefKind::Fn {
+        tcx.sess.span_err(span, "the `proof` attribute can only be applied to functions");
+    } else if tcx.generics_of(def_id).requires_monomorphization(tcx) {
+        tcx.sess.span_err(span, "the `proof` attribute cannot be applied to generic functions");
+    } else {
+        let instance = Instance::mono(tcx, def_id);
+        let helper = CompilerHelpers { tcx };
+        if !helper.fn_abi_of_instance(instance, ty::List::empty()).args.is_empty() {
+            tcx.sess.span_err(span, "functions used as harnesses cannot have any arguments");
```

Comment:
Should this be an error? Is there a way to ignore the arguments and throw a warning? Not sure if that's feasible or not but it doesn't seem like something we should throw an error for.

Reply:
This was already an error. Unfortunately, we can't run harnesses with arguments in a sound way today.